d3a_interface/device_validator.py

```diff
@@ -265,3 +267,85 @@ def validate_storage_device(**kwargs):
+                                     f"{CepSettings.ENERGY_RATE_RANGE.min} & "
+                                     f"{CepSettings.ENERGY_RATE_RANGE.max}."]})
+
+
+def _validate_rate_profile(energy_rate_profile):
+    for date, value in energy_rate_profile.items():
+        print(f"date: {date}")
+        if date == "filename":
+            continue
+        value = float(value) if type(value) == str else value
+        if not CepSettings.ENERGY_RATE_RANGE.min <= value <= CepSettings.ENERGY_RATE_RANGE.max:
+            raise D3ADeviceException(
+                {"misconfiguration": [f"energy_rate should at time: {date} be in between "
+                                      f"{CepSettings.ENERGY_RATE_RANGE.min} & "
+                                      f"{CepSettings.ENERGY_RATE_RANGE.max}."]})
+
+
+def validate_commercial_producer(**kwargs):
+    if "energy_rate" in kwargs and kwargs["energy_rate"] is not None:
+        if type(kwargs["energy_rate"]) is float or type(kwargs["energy_rate"]) is int:
+            _validate_rate(kwargs["energy_rate"])
+        elif type(kwargs["energy_rate"]) is str:
+            _validate_rate_profile(ast.literal_eval(kwargs["energy_rate"]))
+        elif type(kwargs["energy_rate"]) is dict:
+            _validate_rate_profile(kwargs["energy_rate"])
+        else:
+            raise D3ADeviceException({"misconfiguration": [f"energy_rate has an invalid type."]})
+
+
+def validate_market_maker(**kwargs):
+    if "energy_rate" in kwargs and kwargs["energy_rate"] is not None:
+        if type(kwargs["energy_rate"]) is float or type(kwargs["energy_rate"]) is int:
+            _validate_rate(kwargs["energy_rate"])
+        elif type(kwargs["energy_rate"]) is str:
+            _validate_rate_profile(ast.literal_eval(kwargs["energy_rate"]))
+        elif type(ast.literal_eval(kwargs["energy_rate"])) is dict:
+            _validate_rate_profile(kwargs["energy_rate"])
+        else:
+            raise D3ADeviceException({"misconfiguration": [f"energy_rate has an invalid type."]})
+    if "energy_rate_profile" in kwargs and kwargs["energy_rate_profile"] is not None and \
+            ("energy_rate_profile_uuid" not in kwargs or
+             kwargs["energy_rate_profile_uuid"] is None):
+        raise D3ADeviceException(
+            {"misconfiguration": [f"energy_rate_profile must have a uuid."]})
+    if "grid_connected" in kwargs and kwargs["grid_connected"] is not None and \
+            type(kwargs["grid_connected"]) is not bool:
+        raise D3ADeviceException(
+            {"misconfiguration": [f"grid_connected must be a boolean value."]})
+
+
+def validate_finite_diesel_generator(**kwargs):
+    if "max_available_power_kW" in kwargs and kwargs["max_available_power_kW"] is not None:
+        if type(kwargs["max_available_power_kW"]) is float or \
+                type(kwargs["max_available_power_kW"]) is int:
+            if not (CepSettings.MAX_POWER_KW_RANGE.min <= kwargs["max_available_power_kW"] <=
```

Comment:
I know, I should have mentioned this earlier, but a similar check is done a lot of times in the d3a-interface. How about of abstracting this away into a function that only takes the inputs [variable_name, value, min_limit, max_limit]?

Reply:
Sorry, What happened to this comment. Do you plan to implement it?

Reply:
Ah now I see. You only abstracted the min<=x<=max check away. But what i meant was also to put in the raise of the D3ADeviceException in there. That is why I proposed 4 inout variables.